src/serialization/sb2.js

```diff
@@ -809,7 +844,13 @@ const sb2import = function (json, runtime, optForceSprite, zip) {
         extensionIDs: new Set(),
         extensionURLs: new Map()
     };
-    return parseScratchObject(json, runtime, extensions, !optForceSprite, zip)
+    return Promise.resolve(parseScratchAssets(json, runtime, extensions, !optForceSprite, zip))
+        // Force this promise to wait for the next loop in the js tick. Let
+        // storage have some time to send off asset requests.
+        .then(assets => Promise.resolve(assets))
```

Comment:
I missed this (in its other form) the first time around... could you explain a bit more why this is needed? I get that it delays this `then` chain by one tick but why do we need that? It seems like somebody (storage? `parseScratchAssets`?) should return something that resolves at the right time without the need for an arbitrary one-tick delay. Is there a `Promise.all(...)` missing from somewhere maybe?

Reply:
This is a bit of an onion.

ScratchStorage uses promises that delay making the actual requests until later. Some of those promises can return promises themselves, further delaying the actual request. As is we can't know when a request is actually made.

Additionally defining when a request is made I think is tricky. Is it when a fetch or xhr call is made? What if that fails and you try the next tool or source. So then is it when the first tool receives a 200/300 status code? I think may be better to leave this as a vague idea and instead to make part of scratch that are synchronous into green threads.

Activities that take a lot of time like deserializing a project or the initial construction of the gui's dom should be made into green threads where they yield to other work by delaying further work with a promise. For example deserializing an sb2's could yield a promise at the start of each object. That way the stage and each of the sprites would let some other work happen before they do work. Using promises to green thread this (unless those promises used setTimeout or another event) would continue to block the main thread so the render would appear to happen all at once as it does now.

The solution above is a minimal achievement of this idea so that another Issue or PR could have a better discussion of the concept.